# Renaming scheme
> The following scripts were applied to both p10-1 and p10-2 datasets.  

The initial names applied to the **.fastq** files automatically generated by NAU were simplified. The output from NAU for a file followed one of two naming conventions:

- `CONTROL-{SampleName}-xx-EN-USA-2017-076-JF_S{###}_L001_R{#}_001.fastq.gz` for mock community (positive) and negative control samples, with `SampleName` being specific to each sample, `###` being either a two or three digit value assigned to the NAU sample sheet (effectively a redundant sample name they apply to each sample; shared with each forward and reverse **.fastq** file pair), and `#` representing the forward (1) or reverse (2) paired read set.  
- `NHCS-{SampleName}-xx-VE-USA-2017-076-JF_S{###}_L001_R{#}_001.fastq.gz` for all true samples, following the same naming scheme as described above  

The goal was to produce file names with the following scheme: `{SampleName}_{barcode}_L001_R{#}_001.fastq.gz`.

The following steps were applied to the files to achieve that:
> Note that raw files were downloaded to a subdirectory within the main project folder `OahuBird`, thus these raw `.fastq.gz` files all reside in `.../path/to/fqraw`  

Samples will start out looking something like this:
```
CONTROL-asc097-xx-xx-USA-2017-076-JF_S2_L001_R1_001.fastq.gz
CONTROL-asc097-xx-xx-USA-2017-076-JF_S2_L001_R2_001.fastq.gz
...
CONTROL-mockIM4p10L1-xx-EN-USA-2017-076-JF_S1_L001_R1_001.fastq.gz
CONTROL-mockIM4p10L1-xx-EN-USA-2017-076-JF_S1_L001_R2_001.fastq.gz
...
NHCS-asc474-xx-VE-USA-2017-076-JF_S379_L001_R1_001.fastq.gz
NHCS-asc474-xx-VE-USA-2017-076-JF_S379_L001_R2_001.fastq.gz
```

1. Start by stripping off all the parts of the `.fastq.gz` files you don't want.
```
for i in *.gz; do mv "$i" "$(echo "$i" | sed -e 's/NHCS-//' | sed -e 's/CONTROL-//' | sed -e 's/-.*._L/_L/' | sed -e 's/asc/OahuBird-/' - )"; done
```
This should create a list of files that now look something like this:  

```
OahuBird-097_L001_R1_001.fastq.gz
OahuBird-097_L001_R2_001.fastq.gz
...
mockIM4p10L1__L001_R1_001.fastq.gz
mockIM4p10L1__L001_R2_001.fastq.gz
...
OahuBird-474_L001_R1_001.fastq.gz
OahuBird-474_L001_R2_001.fastq.gz
```

2. Create a text file which is just a list of those `fastq.gz` filenames as they currently exist (we'll use this list to build in a file used to substitute one list of names with another):  
```
find . -name "*.gz" | sort | sed -e 's|^\./||' > raw_filelist.txt
cat raw_filelist.txt > tmp_filelist.txt
```

3. Create a barcode list and replace `+` character with `-`. File was obtained from **Run Indexing QC** tabs from [this sheet for p10-1](https://docs.google.com/spreadsheets/d/1nXT6GJNZ3Gy4PHRYa8s9mExXZpY2Y2ayoZTteQsvvsU/edit#gid=0) and [this sheet for p10-2](https://docs.google.com/spreadsheets/d/1ZDOYiIGTKL7_cP8gFDDGn78mRlCs8z-NIUwzzwcnTRw/edit#gid=0). Copy the sample ID and barcode ID's from sheet and paste in new file called `tmp.txt`. Then create a `raw_barcodelist.txt` file to match our sorted list of downloaded .fastq.gz files. Using the new `raw_barcodelist.txt` file, used `sed` to replace the `+` character with a `-` delimiter:
> note you can't copy the entire submitted list - only copy samples with more than 0 reads (NAU doesn't transfer these files but they are still listed in the spreadsheet)  

```
nano tmp.txt
  # paste in two columns of data, then name, save, and quit program
cut tmp.txt -f 1 > tmp1.txt
cut tmp.txt -f 2 > tmp2.txt

cat tmp1.txt | sed -e 's/NHCS-//' | sed -e 's/CONTROL-//' | sed -e 's/-.*.$//' | sed -e 's/asc/OahuBird-/' > tmp3.txt
paste tmp3.txt tmp2.txt > tmp.txt
sort -k1,1 tmp.txt | awk '{print $2}' | sed 's/+/-/' > raw_barcodelist.txt
```

4. Make duplicate of that barcode list, printing same value for a total of two identical strings among two rows.
```
perl -ne 'for$i(0..1){print}' raw_barcodelist.txt > tmp_barcodelist.txt
```

5. Create the left and right edges of the file list you want:
```
cut -d '_' -f 1 tmp_filelist.txt > left.txt
cut -d '_' -f 2-4 tmp_filelist.txt > right.txt
```

6. Now paste it all together to create the final filelist.txt you want to substitute with:
```
paste left.txt tmp_barcodelist.txt right.txt -d "_" > wanted_filelist.txt
```

7. Create a two-column text file containing the initial `tmp_filelist.txt` as the first column (the thing you are substituting), with the `final_filelist.txt` names you'll be replacing with:
```
paste tmp_filelist.txt wanted_filelist.txt > final_filelist.txt
```

8. Use that `final_filelist.txt` and pass this argument to rename all the files:
```
xargs -a final_filelist.txt -n 2 mv
```

This should rename all the files so that they now look something like this (compare to the inital structure from *step 1*:

```
OahuBird-097_CTCGACTT-ATCGTACG_L001_R1_001.fastq.gz
OahuBird-097_CTCGACTT-ATCGTACG_L001_R2_001.fastq.gz
...
mockIM4p10L1_TGCGTCAA-GTCTAGTG_L001_R1_001.fastq.gz
mockIM4p10L1_TGCGTCAA-GTCTAGTG_L001_R2_001.fastq.gz
...
OahuBird-474_CTGCGTAG-GACACCGT_L001_R1_001.fastq.gz
OahuBird-474_CTGCGTAG-GACACCGT_L001_R2_001.fastq.gz
```

Remove the unwanted `.txt` files from the directory and you're ready to start with the `amptk `pipeline:

```
rm *.txt
```


# renaming Cobb raw.fq files for small project
An earlier sequencing effort was undertaken for this project; these samples were not included in the currently described analysis, but were renamed nonetheless as follows:

```
rename -n asc_ asc- *.gz
rename mock_IM3 mock-IM3 *.gz
find . -name "*.gz" | sort | sed -e 's|^\./||' > raw_filelist.txt
cut -d '_' -f 2- raw_filelist.txt > wantedc2.txt
## added wanted metadata names to wantedc0.txt
nano wantedc0.txt ## then pasted wanted metadata names from metadata.csv file
while read line; do for i in {1..2}; do echo "$line"; done; done < wantedc0.txt > wantedc1.txt
paste wantedc1.txt wantedc2.txt -d '_' > wantedlist.txt
paste raw_filelist.txt wantedlist.txt > final_filelist.txt
mv _CGAGAGTT-CGTTACTA_L002_R1_001.fastq.gz mockIM3_CGAGAGTT-CGTTACTA_L002_R1_001.fastq.gz
mv  p_CGAGAGTT-CGTTACTA_L002_R2_001.fastq.gz mockIM3_CGAGAGTT-CGTTACTA_L002_R2_001.fastq.gz
re name NTC_ NTC- *
```
